Gems/Atom/Tools/AtomToolsFramework/Code/Tests/ViewportInteractionImplTests.cpp

```diff
@@ -0,0 +1,171 @@
+    {
+    public:
+        static inline constexpr AzFramework::ViewportId TestViewportId = 1234;
+        static inline AzFramework::ScreenSize ScreenDimensions = AzFramework::ScreenSize(1280, 720);
+
+        static AzFramework::ScreenPoint ScreenCenter()
+        {
+            const auto halfScreenDimensions = ScreenDimensions * 0.5f;
+            return AzFramework::ScreenPoint(halfScreenDimensions.m_width, halfScreenDimensions.m_height);
+        }
+
+        void SetUp() override
+        {
+            AZ::NameDictionary::Create();
+
+            m_view = AZ::RPI::View::CreateView(AZ::Name("TestView"), AZ::RPI::View::UsageCamera);
+
+            const auto aspectRatio = aznumeric_cast<float>(ScreenDimensions.m_width) / aznumeric_cast<float>(ScreenDimensions.m_height);
+
+            AZ::Matrix4x4 viewToClipMatrix;
+            AZ::MakePerspectiveFovMatrixRH(viewToClipMatrix, AZ::DegToRad(60.0f), aspectRatio, 0.1f, 1000.f, true);
+            m_view->SetViewToClipMatrix(viewToClipMatrix);
+
+            m_viewportInteractionImpl = AZStd::make_unique<AtomToolsFramework::ViewportInteractionImpl>(m_view);
+
+            m_viewportInteractionImpl->m_deviceScalingFactorFn = []
+            {
+                return 1.0f;
+            };
+            m_viewportInteractionImpl->m_screenSizeFn = []
+            {
+                return ScreenDimensions;
+            };
+
+            m_viewportInteractionImpl->Connect(TestViewportId);
+        }
+
+        void TearDown() override
+        {
+            m_viewportInteractionImpl->Disconnect();
+            m_viewportInteractionImpl.reset();
+
+            m_view.reset();
+
+            AZ::NameDictionary::Destroy();
+        }
+
+        AZ::RPI::ViewPtr m_view;
+        AZStd::unique_ptr<AtomToolsFramework::ViewportInteractionImpl> m_viewportInteractionImpl;
+    };
+
+    // transform a point from screen space to world space, and then from world space back to screen space
+    AzFramework::ScreenPoint ScreenToWorldToScreen(
+        const AzFramework::ScreenPoint& screenPoint,
+        AzToolsFramework::ViewportInteraction::ViewportInteractionRequests& viewportInteractionRequests)
+    {
+        const auto worldResult = viewportInteractionRequests.ViewportScreenToWorld(screenPoint);
+        return viewportInteractionRequests.ViewportWorldToScreen(worldResult);
+    }
+
+    TEST_F(ViewportInteractionImplFixture, ViewportInteractionRequestsMapsFromScreenToWorldAndBack)
+    {
+        using AzFramework::ScreenPoint;
+
+        m_view->SetCameraTransform(AZ::Matrix3x4::CreateFromMatrix3x3AndTranslation(
+            AZ::Matrix3x3::CreateRotationZ(AZ::DegToRad(90.0f)), AZ::Vector3(10.0f, 0.0f, 5.0f)));
+
+        {
+            const auto expectedScreenPoint = ScreenPoint{ 600, 450 };
+            const auto resultScreenPoint = ScreenToWorldToScreen(expectedScreenPoint, *m_viewportInteractionImpl);
+            EXPECT_EQ(resultScreenPoint, expectedScreenPoint);
```

Comment:
The output of this when the test fails is this:
```
/home/ANT.AMAZON.COM/burelc/ws/scratch/Gems/Atom/Tools/AtomToolsFramework/Code/Tests/ViewportInteractionImplTests.cpp:92: Failure
Expected equality of these values:
  resultScreenPoint
    Which is: 8-byte object <2F-02 00-00 C2-01 00-00>
  expectedScreenPoint
    Which is: 8-byte object <58-02 00-00 C2-01 00-00>
```

We can improve it a bit by adding a function to make a matcher for that `ScreenPoint` type:
```cpp
auto MatchesScreenPoint(const AzFramework::ScreenPoint& screenPoint)
{
    return testing::AllOf(
        testing::Field(&AzFramework::ScreenPoint::m_x, testing::Eq(screenPoint.m_x)),
        testing::Field(&AzFramework::ScreenPoint::m_y, testing::Eq(screenPoint.m_y))
    );
}

EXPECT_THAT(resultScreenPoint, MatchesScreenPoint(expectedScreenPoint));
```

Then the output is:
```
/home/ANT.AMAZON.COM/burelc/ws/scratch/Gems/Atom/Tools/AtomToolsFramework/Code/Tests/ViewportInteractionImplTests.cpp:100: Failure
Value of: resultScreenPoint
Expected: (is an object whose given field is equal to 600) and (is an object whose given field is equal to 450)
  Actual: 8-byte object <2F-02 00-00 C2-01 00-00> (of type AzFramework::ScreenPoint), whose given field is 559 (of type int)
```

We can make that better again by providing a `PrintTo` overload for `ScreenPoint`:
```cpp
namespace AzFramework
{
    void PrintTo(const ScreenPoint& point, std::ostream* os)
    {
        *os << "{x: " << point.m_x << ", y: " << point.m_y << "}";
    }
} // namespace AzFramework
```

Now the output is:
```
/home/ANT.AMAZON.COM/burelc/ws/scratch/Gems/Atom/Tools/AtomToolsFramework/Code/Tests/ViewportInteractionImplTests.cpp:108: Failure
Value of: resultScreenPoint
Expected: (is an object whose given field is equal to 600) and (is an object whose given field is equal to 450)
  Actual: {x: 559, y: 450} (of type AzFramework::ScreenPoint), whose given field is 559 (of type int)
```

I'd advocate for that `PrintTo` function to be in some header inside `AzFramework` itself, similar to how we have that `IsClose` matcher template inside `Code/Framework/AzCore/Tests/AZTestShared/Math/MathTestHelpers.h`

Reply:
Good shout I can add this - so `PrintTo` would be in the header of the type and not in `MathTestHelpers`? As it's testing specific would it not make more sense there?

Reply:
I was thinking some file that only gets included in test targets. I'm not sure if such a file exists for AzFramework currently. I wouldn't put it in the production header, since it's only going to be used in tests.

Reply:
Yup that sounds good, see this commit with the relevant changes - https://github.com/o3de/o3de/pull/5494/commits/f92ee5f58cfe0363ecff10a58686cb0d3f7618a7

Reply:
Output now looks like this for failures